### ISSUE TITLE
GitLab Shared Runner Remote URL Update

### ISSUE THREAD

**Issue by wiebersk** [danger/danger#1231]
# Report

## What did you do?

ℹ Please replace this with what you did.
`danger --fail-on-errors=true --verbose`

## What did you expect to happen?

Danger to run

## What happened instead?

Danger failed with 
```
Could not set up API to Code Review site for Danger
 For Danger to run on this project, you need to expose a set of following the ENV vars:
  - GitHub: DANGER_GITHUB_API_TOKEN
  - GitLab: DANGER_GITLAB_API_TOKEN
  - VSTS: DANGER_VSTS_API_TOKEN, DANGER_VSTS_HOST
  - BitbucketServer: DANGER_BITBUCKETSERVER_USERNAME, DANGER_BITBUCKETSERVER_PASSWORD, DANGER_BITBUCKETSERVER_HOST
  - BitbucketCloud: DANGER_BITBUCKETCLOUD_USERNAME, DANGER_BITBUCKETCLOUD_UUID, DANGER_BITBUCKETCLOUD_PASSWORD
  - LocalOnly: DANGER_LOCAL_ONLY
```

## Your Environment

* Which CI are you running on? GitLab shared runners
* Are you running the latest version of Danger? 8.0.0
* What is your Dangerfile? https://gitlab.com/gitlab-org/gitlab/-/blob/master/Dangerfile

### Other information

GitLab found this at https://gitlab.com/gitlab-org/gitlab/-/issues/219411

TLDR: GitLab Managed Runners remote URL is changed to `ci-api-gprd.gitlab.net` with https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/10069

The workaround that was done at https://gitlab.com/gitlab-org/gitlab/-/issues/219411 was to set `DANGER_GITLAB_HOST` to `ci-api-gprd.gitlab.net`

**Comment by T4cC0re** [comment]
Hey there!

This is the GitLab Engineer working on the abovementioned change.

As mentioned in my comment on https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/10069#note_351787227, from what I can see, danger takes the CI_API_V4_URL to talk to the API https://github.com/danger/danger/blob/master/lib/danger/request_sources/gitlab.rb#L65

This variable is still `gitlab.com`-based, as evidenced here: https://staging.gitlab.com/T4cC0re/test4/-/jobs/37237300#L37 (granted, this is `staging.gitlab.com`, but it is not the clone_url the runner was set to, `ci-api.gstg.gitlab.com`)

So I think, danger might just not detect this properly. IMO it would be cleaner to have `danger` check for `GITLAB_CI` to be `true` ([reference](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html)) and then use the `CI_API_V4_URL`. So I think #1232 might not improve the situation here.

The API host `ci-api-gprd.gitlab.net` does *NOT* represent the full GitLab API, but just a subset and must not be used to communicate with GitLab by tools other than the GitLab CI Runner itself. Every interaction from within GitLab CI jobs with that host is unsupported.

**Comment by T4cC0re** [comment]
Another note:

@wiebersk noted
> Which CI are you running on? GitLab shared runners

While this is correct, these runners are the ones used by GitLab internally. The change was not deployed to public GitLab shared runners, yet.

**Comment by wiebersk** [comment]
> The change was not deployed to public GitLab shared runners, yet.

@T4cC0re - Thank you for adding this context. I thought that that the public GitLab shared runners were included in the change as well. That means the company is only impacted with the change as it stands. 😌 

---

I observed was that the `Danger::RequestSources::RequestSource` was attempting to match `ci-api.gstg.gitlab.com` (`scm.origins`) to `gitlab.com` (`host`) and returned false for `validates_as_ci?` when called by super in `Danger::RequestSources::GitLab#validates_as_ci?`. This led to Danger never identifying a RequestSource.

```
      # @return [Boolean] whether scm.origins is a valid git repository or not
      def validates_as_ci?
        !!self.scm.origins.match(%r{#{Regexp.escape self.host}(:|/)(.+/.+?)(?:\.git)?$})
      end
```

> Does this (`DANGER_GITLAB_HOST`) set the API it talks to, or does this just enable some form of allowlist to enable GitLab integration?

I didn't answer your question from https://gitlab.com/gitlab-com/gl-infra/infrastructure/-/issues/10069#note_351762636

My understanding is `DANGER_GITLAB_HOST` overwrote the host variable for the RequestSource validation above and that the [API client](https://github.com/danger/danger/blob/master/lib/danger/request_sources/gitlab.rb#L38) uses the `endpoint` attribute and `CI_API_V4_URL` to communicate back with GitLab. 

I might be missing something but I saw that the `host` attribute was only used in the comparison to scm.origins: `git remote show origin -n` for `Danger::RequestSources::GitLab`

**Comment by T4cC0re** [comment]
> I thought that that the public GitLab shared runners were included in the change as well.

@wiebersk They are still planned to be included in the change, but the rollout will happen later. I hope we can resolve this issue in a way, that allows us to move forward with the change without asking our users to apply a workaround, but instead just use an updated version of `danger`.

--- 

> My understanding is DANGER_GITLAB_HOST overwrote the host variable for the RequestSource validation above and that the API client uses the endpoint attribute and CI_API_V4_URL to communicate back with GitLab.

If that is indeed the case, the workaround would at least not break anything. :)

**Comment by rymai** [comment]
I opened a new PR at https://github.com/danger/danger/pull/1234.
